packages/logger/index.js

```diff
@@ -0,0 +1,123 @@
+const logger = require('../dd-trace/src/log')
+// path to require tracer logger
+
+const https = require('https')
+const coalesce = require('koalas')
+
+class V2LogWriter {
+  // Note: these attribute names match the corresponding entry in the JSON payload.
+  constructor ({ ddsource, hostname, service, site, apiKey, interval, timeout }) {
+    this.ddsource = ddsource
+    this.hostname = hostname
+    this.service = service
+    this.interval = interval
+    this.timeout = timeout
+    this.buffer = []
+    this.buffer_limit = 1000
+    this.apiKey = apiKey
+    this.endpoint = '/api/v2/logs'
+    this.site = coalesce(site, 'datadoghq.com')
+    this.intake = `http-intake.logs.${this.site}`
+    this.headers = {
+      'DD-API-KEY': this.apiKey,
+      'Content-Type': 'application/json'
+    }
+    this.timer = setInterval(() => {
+      this.flush()
+    }, this.interval)
+
+    logger.debug(`"started log writer to ${this.url}"`)
+  }
+
+  start () {
+    this.timer()
+  }
+
+  tagString (tags) {
+    let tagStr = ''
+    for (const key in tags) {
+      tagStr += key + ':' + tags[key] + ','
+    }
+    return tagStr
+  }
+
+  log (log, tags, span) {
+    const logTags = this.tagString(tags)
+    const toLog = {
+      'timestamp': Date.now(),
+      'message': log.message,
+      'hostname': coalesce(log.hostname, this.hostname),
+      'ddsource': coalesce(log.ddsource, this.ddsource),
+      'service': coalesce(log.service, this.service),
+      'status': log.level,
+      'ddtags': logTags,
+      'dd.trace_id': span.trace_id + '',
+      'dd.span_id': span.span_id + ''
+    }
+    for (const key in log) {
+      if (!toLog[key]) {
+        toLog[key] = log[key]
+      }
+    }
+    return toLog
+  }
+
+  enqueue (log) {
+    if (this.buffer.length >= this.buffer_limit) {
+      logger.warn(`"log buffer full (limit is ${this.buffer_limit}), dropping log"`)
+      this.buffer = []
+      return
+    }
+    this.buffer.push(log)
+  }
+
+  shutdown () {
+    clearInterval(this.timer)
+    this.flush()
+  }
+
+  url () {
+    return `"https://${this.intake}${this.endpoint}"`
+  }
+
+  flush () {
+    let logs
+    let numLogs
+    let encodedLogs
+
+    if (!this.buffer) {
+      return
+    }
+
+    try {
+      logs = this.buffer
+      this.buffer = []
+
+      numLogs = logs.length
+      encodedLogs = JSON.stringify(logs)
+    } catch (error) {
+      logger.error(`"failed to encode ${numLogs} logs"`)
+      return
+    }
+
+    const options = {
+      hostname: this.intake,
+      port: 443,
+      path: this.endpoint,
+      method: 'POST',
+      headers: this.headers
+    }
+    let req
+    try {
```

Comment:
I don't think this try/catch is going to catch an HTTP error. It probably needs to use `req.on('error')` or something like that.

This implementation clears the logs as soon as the flush method is called so there isn't necessarily the need for major error handling. So it might be as simple as something like this:

```js
req.on('error', (e) => {
  logger.error(`failed to send ${numLogs} logs to ${this.intake}, message: ${e.message}`)
})
```

Down the road we could look into more robust solutions like retrying to send logs if a 4XX error happens but that's beyond the current scope.

Reply:
@tlhunter I looked at what errors are grabbed and I do see HTTP errors being caught by the catch, should I still use req.on?

Reply:
Very interesting. Here's some code that I'm using to test with and I can never get the SYNC ERROR part to run:

```js
const https = require('https')

const options = {
  // hostname: 'thomashunter.name',
  hostname: 'FAKEDOMAIN.thomashunter.name',
  port: 443,
  path: '/',
  method: 'GET',
  headers: {}
}

try {
  const req = https.request(options, (res) => {
    console.log(`ASYNC SUCCESS: ${res.statusCode}`)
  })

  req.on('error', (e) => {
    console.error('ASYNC ERR:', e.message)
  })

  req.end()
} catch (err) {
  console.log('SYNC ERROR:', err)
}
```

But you're seeing the `failed to send ${numLogs} logs to ${this.intake}` message get displayed when you run the code? What is the `error.message` when this happens?

Reply:
I am not seeing it caught anymore, as well as 400 responses aren't currently being handled. So I have to add something to handle those responses. `req.on` doesn't handle the 400 responses.

Reply:
Ok, so it looks like the way to proceed with this is to use `req.on` to handle the async errors and anything that is returned with. response higher than 300 we can use `req.once` and this way we an check the res status code on the response object after the request has finished. I am thinking this could be the way we handle synchronous errors 🤷🏼‍♀️

Reply:
My request looks like this
```
const req = https.request(options, (res) => {
      tracerLogger.info(`statusCode: ${res.statusCode}`)
      res.on('data', function (d) {
        console.log('res', d)
      })
    })
    req.on('error', (e) => {
      console.error('error', e.message)
    })
    req.write(encodedLogs)
    req.end()
    req.once('response', (res) => {
      if (res.statusCode >= 300) {
        tracerLogger.error(`failed to send ${numLogs} logs to ${this.intake}`)
      }
    })
```